### DISCUSSION TITLE
Can a project using react-native-tvos also run fine on iOS/iPad?

### DISCUSSION THREAD

**Discussion by adamkoch** [react-native-tvos/react-native-tvos#645] [Q&A] [answered]
Just wondering if there's any reason a project using react-native-tvos will also be able to run on iOS/iPadOS assuming the correct configuration is set up in the Xcode project?

Background: I'm still on a much older version of RN (0.61.5) before TV was split out. I'm wondering what my options are for upgrading to latest RN. I primarily run TV apps (AndroidTV and AppleTV) but also run on iPad. So if I switch from rn 0.61.5 to latest react-native-tv will it break my iPad build?

Thanks!

**Comment by douglowder** [comment, marked as answer]
Yes this repo supports building for both iOS/iPad and tvOS. It also supports building for Android phones and tablets in addition to Android TV.

This is most easily accomplished by creating an Expo project for TV (https://docs.expo.dev/guides/building-for-tv/). Projects created this way have their native code generated dynamically during the Expo prebuild step. By setting (or not setting) the EXPO_TV environment variable, the native code generates a TV app (or a phone app).

By following the above process, you will be able to build (separately) for iPhone/iPad, and for Apple TV.

In older versions of RN like 0.61.5, the project Podfile contained targets for both iOS and tvOS. That is no longer supported, due to some complexities in the include paths used for including generated code. So you can certainly build for both targets, but they require separate builds with different Podfiles that have only one target each. The Expo process described above automatically generates the correct Podfile for whichever platform you choose to build.

**Reply by adamkoch** [reply]
Fantastic, thanks for the reply. What about without Expo? As some of the libraries we use are not Expo compatible.

**Reply by douglowder** [reply]
Changing from iOS to tvOS build without Expo would require modifying the Podfile to change the target platform, deleting ios/build and ios/Pods, and rerunning pod install.

I was not aware that there were libraries incompatible with Expo... what are they?

**Reply by adamkoch** [reply]
Ah well maybe my knowledge is out of date here - but we had several libraries that used rn linking such as react-native-fs and react-native-video and at the time those were not compatible with expo. However, a quick google sees to show react-native-video likely works now but react-native-fs may not (there is expo-file-system instead, not sure how it compares).

**Reply by douglowder** [reply]
I had a quick look at react-native-fs, it might work but does not appear to have had any maintenance recently. I think that if you decide to go with an Expo app, you would be much better off with expo-file-system (more features, promise-based APIs, etc.). Just don't try to use `documentsDirectory` on Apple TV 😄 For video, I would recommend `expo-av` (https://docs.expo.dev/versions/latest/sdk/video/). Here is an example of how to use it on TV: https://github.com/douglowder/VideoTest/blob/main/App.tsx

**Reply by adamkoch** [reply]
Amazing, thank you. Will be trying this out shortly and see if I can upgrade and migrate our code over to react-native-tvos and expo. I really appreciate the helpful replies and info!
